core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1078,7 +1078,9 @@ abstract class RDD[T: ClassTag](
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
       // If creating an extra level doesn't help reduce
       // the wall-clock time, we stop tree aggregation.
-      while (numPartitions > scale + numPartitions / scale) {
+
+      // Don't trigger TreeAggregation when it doesn't save wall-clock time
+      while (numPartitions > scale + math.ceil((numPartitions).toDouble / scale)) {
```

Comment:
LGTM except the parens around `(numPartitions)` aren't needed. That's trivial though.